# Bismark

Bismark is a program to map bisulfite treated sequencing reads to a genome of interest and perform methylation calls in a single step. The output can be easily imported into a genome viewer, such as [SeqMonk](http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/), and enables a researcher to analyse the methylation levels of their samples straight away. It's main features are:

* Bisulfite mapping and methylation calling in one single step
* Supports single-end and paired-end read alignments
* Supports ungapped and gapped alignments
* Alignment seed length, number of mismatches etc. are adjustable
* Output discriminates between cytosine methylation in `CpG`, `CHG` and `CHH` `*` context

This link will take you to the [Bismark](http://www.ncbi.nlm.nih.gov/pubmed/21493656) publication.

This link will take you to our review about primary data analysis in [BS-Seq](http://www.ncbi.nlm.nih.gov/pubmed/22290186).

This link will take you to our protocol '[Quality Control, trimming and alignment of Bisulfite-Seq data](http://www.epigenesys.eu/en/protocols/bio-informatics/483-quality-control-trimming-and-alignment-of-bisulfite-seq-data-prot-57)' at the Epigenesys website.

Here you can access the Bismark documentation [Bismark User Guide](http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf) (pdf).

Here is an overview of the alignment modes that are currently supported by Bismark: [Bismark alignment modes](http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_alignment_modes.pdf) (pdf).
